deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/ocnn/OCNNOutputLayer.java

```diff
@@ -0,0 +1,244 @@
+
+        return new Pair<>(pair.getFirst(), epsilonNext);
+    }
+
+
+    /** Returns tuple: {Grafdient,Delta,Output} given preOut */
+    private Pair<Gradient, INDArray> getGradientsAndDelta(INDArray preOut, LayerWorkspaceMgr workspaceMgr) {
+        ILossFunction lossFunction = layerConf().getLossFn();
+        INDArray labels2d = getLabels2d(workspaceMgr, ArrayType.BP_WORKING_MEM);
+        INDArray delta = lossFunction.computeGradient(labels2d, preOut, layerConf().getActivationFn(), maskArray);
+
+        Gradient gradient = new DefaultGradient();
+        INDArray vGradView = gradientViews.get(V_KEY);
+        double oneDivNu = 1.0 / layerConf().getNu();
+        INDArray xTimesV = input.mmul(getParam(V_KEY));
+        INDArray derivW = layerConf().getActivationFn()
+                .getActivation(xTimesV.dup(),true).negi()
+                .muliRowVector(delta).mean(0)
+                .muli(oneDivNu).addi(getParam(W_KEY));
+        gradient.setGradientFor(W_KEY,gradientViews.get(W_KEY).assign(derivW));
+
+        //dG -> sigmoid derivative
+
+        INDArray firstVertDerivV =  layerConf().getActivationFn()
+                .backprop(xTimesV.dup(),Nd4j.scalar(1.0))
+                .getFirst().muliRowVector(getParam(W_KEY).neg())
+                .muliRowVector(delta)
+                .reshape('f',input.size(0),1,layerConf().getHiddenSize());
+        INDArray secondTermDerivV = input.reshape('f',
+                input.size(0),getParam(V_KEY).size(0),1);
+
+        int[]  shape = new int[firstVertDerivV.shape().length];
+        for(int i = 0; i < firstVertDerivV.rank(); i++) {
+            shape[i] = Math.max(firstVertDerivV.size(i),secondTermDerivV.size(i));
+        }
+
+        INDArray mulResult = firstVertDerivV.broadcast(shape)
+                .muli(secondTermDerivV.broadcast(shape));
+
+        INDArray derivV = mulResult
+                .mean(0).muli(oneDivNu).addi(getParam(V_KEY));
+        gradient.setGradientFor(V_KEY,vGradView.assign(derivV));
+
+        INDArray derivR = Nd4j.scalar(delta.meanNumber()).muli(oneDivNu).addi(-1);
+        gradient.setGradientFor(R_KEY,Nd4j.scalar(0.0));
+
+        gradient.setGradientFor(R_KEY,gradientViews.get(R_KEY).assign(derivR));
+        clearNoiseWeightParams();
+
+        return new Pair<>(gradient, delta);
+    }
+
+    @Override
+    public INDArray activate(INDArray input, boolean training, LayerWorkspaceMgr workspaceMgr) {
+        this.input = input;
+        return doOutput(training,workspaceMgr);
+    }
+
+    /**{@inheritDoc}
+     */
+    @Override
+    public double f1Score(INDArray examples, INDArray labels) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public INDArray labelProbabilities(INDArray examples) {
+        //LayerWorkspaceMgr.noWorkspacesImmutable()
+        this.input = examples;
+        return doOutput(false,LayerWorkspaceMgr.noWorkspacesImmutable()).subi(getParam(R_KEY));
+    }
+
+    public INDArray getInput() {
```

Comment:
Unnecessary method given AbstractLayer ```@Data``` annotation?
(And if we do keep it - add override annotation)

Reply:
fixed